dashboard/lti_new.py

```diff
@@ -0,0 +1,197 @@
+                                            last_name=last_name)
+    user_obj.backend = 'django.contrib.auth.backends.ModelBackend'
+    django.contrib.auth.login(request, user_obj)
+    user_roles = course_user_roles(roles, username)
+    is_instructor = check_if_instructor(user_roles, username, course_id)
+
+    course_details = None
+    is_course_data_loaded = False
+
+    try:
+        course_details = Course.objects.get(canvas_id=course_id)
+    except ObjectDoesNotExist:
+        if is_instructor or user_obj.is_staff:
+            canvas_long_id = canvas_id_to_incremented_id(course_id)
+            course_details = Course.objects.create(id=canvas_long_id, canvas_id=course_id, name=course_name)
+            CourseViewOption.objects.create(course_id=canvas_long_id)
+
+    if course_details is None:
+        logger.info(f'Course {course_id} data has not yet been loaded')
+    elif course_details.term_id is not None:
+        logger.info(f'Course {course_id} data is ready')
+        is_course_data_loaded = True
+
+    myla_globals = {
+        "username": username,
+        "is_superuser": user_obj.is_staff,
+        "login": settings.LOGIN_URL,
+        "logout": settings.LOGOUT_URL,
+        "user_image": user_img,
+        "primary_ui_color": settings.PRIMARY_UI_COLOR,
+        "help_url": settings.HELP_URL,
+        "google_analytics_id": settings.GA_ID,
+        "user_courses_info": [
+            {"course_id": course_id, "course_name": course_name, "enrollment_type": short_user_role_list(user_roles)}],
+        "lti_launch_id": launch_id,
+        "lti_is_course_data_loaded": is_course_data_loaded,
+    }
+    return myla_globals
+
+
+@csrf_exempt
+def login(request):
+    try:
+        tool_conf = get_tool_conf()
+    except (OSError, Exception) as e:
+        error_message = {
+            'lti_error_message': f'LTI Login failed due to {e}'
+        }
+        return JsonResponse(error_message, status=500)
+
+    target_link_uri = request.POST.get('target_link_uri', request.GET.get('target_link_uri'))
+    if not target_link_uri:
+        error_message = {
+            'lti_error_message': 'LTI Login failed due to missing "target_link_uri" param'
+        }
+        return JsonResponse(error_message, status=500)
+    dummy_cache = DUMMY_CACHE in settings.DB_CACHE_CONFIGS['BACKEND']
+    launch_data_storage = DjangoCacheDataStorage(cache_name='default') if not dummy_cache else None
+    oidc_login = DjangoOIDCLogin(request, tool_conf, launch_data_storage=launch_data_storage)
+    return oidc_login.enable_check_cookies().redirect(target_link_uri)
+
+
+@require_POST
+@csrf_exempt
+def launch(request):
+    try:
+        tool_conf = get_tool_conf()
+    except (OSError, Exception) as e:
+        error_message = {
+            'lti_error_message': f'LTI Launch failed due to {e}'
+        }
+        return JsonResponse(error_message, status=500)
```

Comment:
This repeated code appears in both `login()` and `launch()` and can be refactored into a new function.